File: 2/2_cuatrimestre/PDOO/practicas/practica 3/deepSpace_3/src/deepspace/Hangar.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author manbolq
 */
public class Hangar {
    private int maxElements;
    private ArrayList<ShieldBooster> shieldBoosters;
    private ArrayList<Weapon> weapons;
    
    
    
    /**************************************************************************/
    /*************************    CONSTRUCTORES   *****************************/
    /**************************************************************************/
    
    Hangar(int capacity){
        maxElements = capacity;
        shieldBoosters = new ArrayList<>(0);
        weapons = new ArrayList<>(0);
    }
    
    Hangar(Hangar h){
        maxElements = h.maxElements;
        shieldBoosters = h.shieldBoosters;
        weapons = h.weapons;
    }
    
    
    /**************************************************************************/
    /****************************    VERSION   ********************************/
    /**************************************************************************/
    
    HangarToUI getUIversion(){
        return new HangarToUI(this);
    }
    
    
    /**************************************************************************/
    /****************************    METODOS   ********************************/
    /**************************************************************************/
    
    private boolean spaceAvailable(){
        return (shieldBoosters.size() + weapons.size() < maxElements);
    }
    
    
    public boolean addWeapon(Weapon w){
        if (spaceAvailable()){
            weapons.add(w);
            return true;
        }
        else return false;
    }
    
    
    public boolean addShieldBooster(ShieldBooster s){
        if (spaceAvailable()){
            shieldBoosters.add(s);
            return true;
        }
        else return false;
    }
    
    
    public int getMaxElements(){
        return maxElements;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters(){
        return shieldBoosters;
    }
    
    public ArrayList<Weapon> getWeapons(){
        return weapons;
    }
    
    public ShieldBooster removeShieldBooster(int s){
        if (s < shieldBoosters.size() && s >= 0){
            ShieldBooster aux = new ShieldBooster(shieldBoosters.get(s));
            shieldBoosters.remove(s);
            return aux;
        }
        else return null;
    }
    
    
    public Weapon removeWeapon(int w){
        if (w < weapons.size() && w >= 0){
            Weapon aux = new Weapon(weapons.get(w));
            weapons.remove(w);
            return aux;
        }
        else return null;            
    }
    
    
    @Override
    public String toString(){
        return getUIversion().toString();
    }
    
}

File: 2/2_cuatrimestre/PDOO/practicas/practica 2/deepSpaceRuby/Hangar.rb
require_relative 'ShieldBooster'
require_relative 'Weapon'
require_relative 'HangarToUI'


module Deepspace
    class Hangar

        attr_reader :maxElements, :weapons, :shieldBoosters
        
        def initialize(capacity)
            @maxElements = capacity
            @weapons = Array.new
            @shieldBoosters = Array.new
        end

        def self.newCopy(h)
            new(h.maxElements)
            if (h.weapons != nil)
                @weapons = h.weapons
            else
                @weapons = Array.new
            end

            if (h.shieldBoosters != nil)
                @shieldBoosters = h.shieldBoosters
            else
                @shieldBoosters = Array.new
            end
        end

        def getUIversion
            return HangarToUI.new(self)
        end

        def addWeapon(w)
            if spaceAvailable
                @weapons << w
                return true
            else 
                false
            end
        end

        def addShieldBooster(s)
            if spaceAvailable
                @shieldBoosters << s
                return true
            else 
                false
            end
        end

        def removeShieldBooster(s)
            if (s < @shieldBoosters.size && s >= 0)
                aux = ShieldBooster.newCopy(@shieldBoosters[s])
                @shieldBoosters.delete_at(s)
                aux
            else
                nil
            end

        end

        def removeWeapon(w)
            if (w < @weapons.size && w >= 0)
                aux = Weapon.newCopy(@weapons[w])
                @weapons.delete_at(w)
                aux
            else
                nil
            end
        end

        def to_s
            return getUIversion.to_s
        end
        
        def spaceAvailable
            @weapons.length + @shieldBoosters.length < @maxElements
        end

    end
end 